From 83c465fffad9144c32025c347bebd0904c0f10b6 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Jonas=20=C3=85dahl?= <jadahl@gmail.com>
Date: Mon, 14 Sep 2026 16:27:40 +0200
Subject: [PATCH] monitor-manager: Move power save setting into helper

This will be useful for tests, which want to set the power save mode
without having to go via D-Bus.

Part-of: <https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/5324>
(cherry picked from commit 9f14039b20a1cb8179b4552291b290cf204ad027)
---
 src/backends/meta-monitor-manager-private.h |  4 ++++
 src/backends/meta-monitor-manager.c         | 26 ++++++++++++++-------
 2 files changed, 22 insertions(+), 8 deletions(-)

diff --git a/src/backends/meta-monitor-manager-private.h b/src/backends/meta-monitor-manager-private.h
index 7b45fb8b49..2e7ff97e32 100644
--- a/src/backends/meta-monitor-manager-private.h
+++ b/src/backends/meta-monitor-manager-private.h
@@ -289,6 +289,10 @@ void                meta_monitor_manager_get_screen_size   (MetaMonitorManager *
 
 MetaPowerSave       meta_monitor_manager_get_power_save_mode (MetaMonitorManager *manager);
 
+META_EXPORT_TEST
+void                meta_monitor_manager_set_power_save_mode (MetaMonitorManager *monitor_manager,
+                                                              MetaPowerSave       mode);
+
 void                meta_monitor_manager_power_save_mode_changed (MetaMonitorManager        *manager,
                                                                   MetaPowerSave              mode,
                                                                   MetaPowerSaveChangeReason  reason);
diff --git a/src/backends/meta-monitor-manager.c b/src/backends/meta-monitor-manager.c
index bee591b7a3..0534c27870 100644
--- a/src/backends/meta-monitor-manager.c
+++ b/src/backends/meta-monitor-manager.c
@@ -312,6 +312,23 @@ meta_monitor_manager_power_save_mode_changed (MetaMonitorManager        *manager
     meta_orientation_manager_uninhibit_tracking (orientation_manager);
 }
 
+void
+meta_monitor_manager_set_power_save_mode (MetaMonitorManager *monitor_manager,
+                                          MetaPowerSave       mode)
+{
+  MetaPowerSaveChangeReason reason;
+  MetaMonitorManagerClass *klass;
+
+  klass = META_MONITOR_MANAGER_GET_CLASS (monitor_manager);
+  if (klass->set_power_save_mode)
+    klass->set_power_save_mode (monitor_manager, mode);
+
+  reason = META_POWER_SAVE_CHANGE_REASON_MODE_CHANGE;
+  meta_monitor_manager_power_save_mode_changed (monitor_manager,
+                                                mode,
+                                                reason);
+}
+
 static void
 power_save_mode_changed (MetaMonitorManager *manager,
                          GParamSpec         *pspec,
@@ -319,9 +336,7 @@ power_save_mode_changed (MetaMonitorManager *manager,
 {
   MetaMonitorManagerPrivate *priv =
     meta_monitor_manager_get_instance_private (manager);
-  MetaMonitorManagerClass *klass;
   int mode = meta_dbus_display_config_get_power_save_mode (manager->display_config);
-  MetaPowerSaveChangeReason reason;
 
   if (mode == META_POWER_SAVE_UNSUPPORTED)
     return;
@@ -333,12 +348,7 @@ power_save_mode_changed (MetaMonitorManager *manager,
       return;
     }
 
-  klass = META_MONITOR_MANAGER_GET_CLASS (manager);
-  if (klass->set_power_save_mode)
-    klass->set_power_save_mode (manager, mode);
-
-  reason = META_POWER_SAVE_CHANGE_REASON_MODE_CHANGE;
-  meta_monitor_manager_power_save_mode_changed (manager, mode, reason);
+  meta_monitor_manager_set_power_save_mode (manager, mode);
 }
 
 void
-- 
2.53.0

